### PR TITLE
[Performance] Performance improvements in non-blockwise fp8 CUTLASS MoE

### PR DESCRIPTION
**Pull request by ElizaWszola**
This PR introduces a couple performance improvements to non-blockwise fp8 CUTLASS MoE. The improvements are:
- Bringing back pre-calculation of `ab_strides` and `c_strides`. It had been disabled due to PPLX codepath compatibility issues which are now resolved.
- Faster kernels for shuffling `hidden_states`, input scales and outputs of the function that executes CUTLASS MoE layer.

#### Benchmarks

Execution times of function`cutlass_moe_fp8` times in microseconds (μs):

```
-------------------------------------------------------------------------------------------------
                                                | current grouped_gemm_moe | new grouped_gemm_moe
-------------------------------------------------------------------------------------------------
num_experts=8, topk=2, MKN=((1, 4096, 28672))   | 8480.9                   | 8143.1
num_experts=8, topk=2, MKN=((4, 4096, 28672))   | 17199.9                  | 16930.2
num_experts=8, topk=2, MKN=((8, 4096, 28672))   | 26038.6                  | 25675.8
num_experts=8, topk=2, MKN=((16, 4096, 28672))  | 26167.1                  | 25848.8
num_experts=8, topk=2, MKN=((32, 4096, 28672))  | 26402.4                  | 26026.9
num_experts=8, topk=2, MKN=((64, 4096, 28672))  | 26816.0                  | 26389.1
num_experts=8, topk=2, MKN=((128, 4096, 28672)) | 27674.2                  | 27175.4
num_experts=8, topk=2, MKN=((256, 4096, 28672)) | 30751.2                  | 29755.0
num_experts=8, topk=2, MKN=((512, 4096, 28672)) | 38974.5                  | 38203.2
num_experts=8, topk=2, MKN=((1, 14336, 4096))   | 5662.6                   | 5385.5
num_experts=8, topk=2, MKN=((4, 14336, 4096))   | 11083.1                  | 10783.2
num_experts=8, topk=2, MKN=((8, 14336, 4096))   | 12398.0                  | 13571.3
num_experts=8, topk=2, MKN=((16, 14336, 4096))  | 11243.3                  | 13677.9
num_experts=8, topk=2, MKN=((32, 14336, 4096))  | 14217.7                  | 13805.0
num_experts=8, topk=2, MKN=((64, 14336, 4096))  | 14595.0                  | 14073.9
num_experts=8, topk=2, MKN=((128, 14336, 4096)) | 15392.0                  | 14538.9
num_experts=8, topk=2, MKN=((256, 14336, 4096)) | 17839.9                  | 16317.0
num_experts=8, topk=2, MKN=((512, 14336, 4096)) | 25584.6                  | 21823.1
num_experts=64, topk=6, MKN=((1, 2048, 1408))   | 2365.9                   | 2013.4
num_experts=64, topk=6, MKN=((4, 2048, 1408))   | 3615.8                   | 3224.5
num_experts=64, topk=6, MKN=((8, 2048, 1408))   | 4391.0                   | 4359.8
num_experts=64, topk=6, MKN=((16, 2048, 1408))  | 5686.3                   | 5189.1
num_experts=64, topk=6, MKN=((32, 2048, 1408))  | 6670.1                   | 6260.9
num_experts=64, topk=6, MKN=((64, 2048, 1408))  | 6955.2                   | 6528.7
num_experts=64, topk=6, MKN=((128, 2048, 1408)) | 7538.7                   | 6893.7
num_experts=64, topk=6, MKN=((256, 2048, 1408)) | 8498.4                   | 7582.4
num_experts=64, topk=6, MKN=((512, 2048, 1408)) | 10094.3                  | 8767.7
num_experts=32, topk=8, MKN=((1, 1024, 1024))   | 1838.5                   | 1531.2
num_experts=32, topk=8, MKN=((4, 1024, 1024))   | 2479.6                   | 2121.3
num_experts=32, topk=8, MKN=((8, 1024, 1024))   | 2730.8                   | 2351.4
num_experts=32, topk=8, MKN=((16, 1024, 1024))  | 2758.8                   | 2388.1
num_experts=32, topk=8, MKN=((32, 1024, 1024))  | 2818.1                   | 2445.5
num_experts=32, topk=8, MKN=((64, 1024, 1024))  | 2942.6                   | 2534.9
num_experts=32, topk=8, MKN=((128, 1024, 1024)) | 3316.4                   | 2745.9
num_experts=32, topk=8, MKN=((256, 1024, 1024)) | 3991.2                   | 3238.2
num_experts=32, topk=8, MKN=((512, 1024, 1024)) | 5357.5                   | 4314.8
num_experts=40, topk=8, MKN=((1, 1024, 1536))   | 1988.2                   | 1787.0
num_experts=40, topk=8, MKN=((4, 1024, 1536))   | 2928.6                   | 2461.8
num_experts=40, topk=8, MKN=((8, 1024, 1536))   | 3151.5                   | 2930.3
num_experts=40, topk=8, MKN=((16, 1024, 1536))  | 3611.4                   | 3209.7
num_experts=40, topk=8, MKN=((32, 1024, 1536))  | 3687.8                   | 3310.0
num_experts=40, topk=8, MKN=((64, 1024, 1536))  | 3846.8                   | 3438.9
num_experts=40, topk=8, MKN=((128, 1024, 1536)) | 4267.1                   | 3689.6
num_experts=40, topk=8, MKN=((256, 1024, 1536)) | 4986.2                   | 4230.9
num_experts=40, topk=8, MKN=((512, 1024, 1536)) | 6593.6                   | 5527.7
```

### CODE REVIEW

File: tests/kernels/moe/test_cutlass_moe.py [hunk 1]
```diff
@@ -439,6 +443,11 @@ def test_run_cutlass_moe_fp8(
         expert_map[start:end] = list(range(num_local_experts))
         expert_map = torch.tensor(expert_map, dtype=torch.int32, device="cuda")
 
+        ab_strides1 = torch.full((e, ), k, device="cuda", dtype=torch.int64)
+        ab_strides2 = torch.full((e, ), n, device="cuda", dtype=torch.int64)
+        c_strides1 = torch.full((e, ), 2 * n, device="cuda", dtype=torch.int64)
+        c_strides2 = torch.full((e, ), k, device="cuda", dtype=torch.int64)
```
Comment (tlrmchlsmth) [inline tests/kernels/moe/test_cutlass_moe.py:449]: Could we save on memory by using the same tensor for both `ab_strides1` and `cstrides2`?

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 2]
```diff
@@ -152,27 +155,11 @@ def run_cutlass_moe_fp8(
                                     problem_sizes1, problem_sizes2, a_map,
                                     c_map, global_num_experts, N, K)
 
-        a1q = _fp8_perm(a1q, a_map)
-        a1q_scale = a1q_scale[a_map] if per_act_token else a1q_scale
+        a1q = ops.shuffle_rows(a1q, a_map)
+        a1q_scale = (ops.shuffle_rows(a1q_scale, a_map)
+                     if per_act_token else a1q_scale)
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py:160]: Does the speedup mentioned in this statement
>Faster kernels for shuffling hidden_states, input scales and outputs of the function that executes CUTLASS MoE layer.

come from being able to use `shuffle_rows` when the `num_cols % 128 == 0` condition is not guaranteed?

**Review by tlrmchlsmth — APPROVED** [review]

File: tests/kernels/moe/test_cutlass_moe.py (hunk 1 above)
Comment (ElizaWszola) [reply tests/kernels/moe/test_cutlass_moe.py:449]: Yes, this should be possible. I'll push an update

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py (hunk 2 above)
Comment (ElizaWszola) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:160]: It comes from using these custom kernels instead of the pytorch index-based function. The "slow" kernels are expected to only be run to shuffle scales.
